core/templates/dev/head/services/SpeechSynthesisChunkerService.js

```diff
@@ -0,0 +1,156 @@
+      else {
+        var chunkLength = (settings && settings.chunkLength) || 160;
+        var pattRegex = new RegExp('^[\\s\\S]{' +
+          Math.floor(chunkLength / 4) + ',' + chunkLength + '}[' +
+          PUNCTUATION_MARKS_TO_END_CHUNKS + ']{1}|^[\\s\\S]{1,' +
+          chunkLength + '}$|^[\\s\\S]{1,' + chunkLength + '} ');
+        var chunkArr = txt.match(pattRegex);
+        if (chunkArr == null ||
+            chunkArr[0] === undefined ||
+            chunkArr[0].length <= 2) {
+          // Call once all text has been spoken.
+          if (callback !== undefined) {
+            callback();
+          }
+          return;
+        }
+        var chunk = chunkArr[0];
+        newUtt = new SpeechSynthesisUtterance(chunk);
+        var x;
+        for (x in utt) {
+          if (x !== 'text') {
+            newUtt[x] = utt[x];
+          }
+        }
+        newUtt.onend = function() {
+          if (_speechUtteranceChunker.cancel) {
+            _speechUtteranceChunker.cancel = false;
+            return;
+          }
+          settings.offset = settings.offset || 0;
+          settings.offset += chunk.length;
+          _speechUtteranceChunker(utt, settings, callback);
+        }
+      }
+   
+      if (settings.modifier) {
+        settings.modifier(newUtt);
+      }
+
+      // IMPORTANT!! Do not remove: Logging the object out fixes some onend
+      // firing issues. Placing the speak invocation inside a callback
+      // fixes ordering and onend issues.
+      console.log(newUtt);
+      $timeout(function () {
+        speechSynthesis.speak(newUtt);
+      }, 0);
+    };
+
+    var _preprocessHtmlForChunking = function(html) {
+      // Replace certain HTML elements with periods to indicate
+      // pauses in speaking. Also, for some reason, there's a lot
+      // of whitespace (like hundreds of characters) so we trim
+      // it off to avoid blank chunks.
+      html = html.replace(new RegExp('</li>', 'g'), '.').trim();
+      // Strip away HTML tags.
+      var tmp = document.createElement('div');
+      tmp.innerHTML = html;
+      var textToSpeak = tmp.textContent || tmp.innerText || '';
+
+      // Insert a space after punctuation marks to ensure that chunking will
+      // end on the desired punctuation marks so that SpeechSynthesis will
+      // pause more naturally. Remove any punctuation marks that have no
+      // effect on speaking.
+      for (var i = 0; i < textToSpeak.length; i++) {
+        if (PUNCTUATION_MARKS_TO_END_CHUNKS.indexOf(
+            textToSpeak.charAt(i)) > -1) {
+          textToSpeak = _insertSpaceAfterIndex(textToSpeak, i);
+          i++;
+        }
+        if (PUNCTUATION_MARKS_TO_IGNORE.indexOf(
+            textToSpeak.charAt(i)) > -1) {
+          textToSpeak = _removeCharacterAtIndex(textToSpeak, i);
+        }
+      }
+
+      return textToSpeak;
+    };
+
+    var _insertSpaceAfterIndex = function(targetString, index) {
+      return targetString.substr(0, index + 1) + ' ' +
```

Comment:
So much copying and creation of new strings ... perhaps just build up the new string character by character in the for loop instead?

Reply:
done